File: tools_C++/src/TestCase.cpp
/*
 * TestCase.cpp
 *
 *  Created on: Apr 13, 2017
 *      Author: broihier
 */

#include "TestCase.h"
#include <iostream>
#include <stdio.h>


TestCase::TestCase(std::string id, std::string category, TestDatabase& database) {
	std::cout << "inside of test case constructor" << std::endl;
	DOMizedTemplate = new DOMizeTemplate("../tools_configuration/TEST_CASE_TEMPLATE.html","./"+category+"/"+id+".html");
	const xml_document& doc = DOMizedTemplate->getDocument();
	status = "UNTESTED";
	xpath_node field = doc.select_node("//td[@id='testCaseId']");
	if (field) {
	    field.node().append_child(node_pcdata).set_value(id.c_str());
	} else {
		std::cout << "testCaseId not found" << std::endl;
	}

	field = doc.select_node("//td[@id='title']");
	if (field) {
	    field.node().append_child(node_pcdata).set_value(database.getIdentificationText(id).c_str());
	} else {
		std::cout << "title not found" << std::endl;
	}

	field = doc.select_node("//ul[@id='requirementsList']");
	if (field) {
		for (std::string requirementId : Utilities::split(database.getRequirementsInThisTest(id),",")) {
			std::string trimmedRequirementId = Utilities::trim(requirementId);
			xml_node listItem = field.node().append_child("li");
			listItem.append_child(node_pcdata).set_value(database.getRequirementsText(trimmedRequirementId).c_str());
		}
	} else {
		std::cout << "requirementsList not found" << std::endl;
	}

	field = doc.select_node("//td[@id='objective']");
	if (field) {
		xml_node paragraphItem = field.node().insert_child_before("p",field.node().first_child());
		paragraphItem.append_child(node_pcdata).set_value("This test case is intended to verify the following implied or specified requirements:");
		paragraphItem = field.node().append_child("p");
		paragraphItem.append_child(node_pcdata).set_value(database.getObjectiveText(id).c_str());
	} else {
		std::cout << "objective not found" << std::endl;
	}

	field = doc.select_node("//ul[@id='setup']");
	if (field) {
		for (std::string setupItem : Utilities::split(database.getSetupText(id),"/hr")) {
			xml_node listItem = field.node().append_child("li");
			listItem.append_child(node_pcdata).set_value(setupItem.c_str());
		}
	} else {
		std::cout << "setup not found" << std::endl;
	}

	field = doc.select_node("//ol[@id='actions']");
	if (field) {
		for (std::string setupItem : Utilities::split(database.getProcedureText(id),"/step")) {
			if (setupItem == "") continue;
			bool procedureStep = true;
			bool firstBullet = true;
			xml_node unOrderedList;
			for (std::string item : Utilities::split(setupItem,"/bullet")) {
				if (procedureStep) {
			        xml_node listItem = field.node().append_child("li");
			        if (item.find("\\step") != std::string::npos) {
			            item.replace(item.find("\\step"),item.find("\\step")+strlen("\\step"),"");
			        }
			        listItem.append_child(node_pcdata).set_value(item.c_str());
			        procedureStep = false;
				} else {
					if (firstBullet) {
					    unOrderedList = field.node().append_child("ul");
					    firstBullet = false;
					}
					if (item.find("\\step") != std::string::npos) {
			            item.replace(item.find("\\step"),item.find("\\step")+strlen("\\step"),"");
					}
					if (item.find("\\bullet") != std::string::npos) {
			            item.replace(item.find("\\bullet"),item.find("\\bullet")+strlen("\\bullet"),"");
					}
			        unOrderedList.append_child("li");
					unOrderedList.append_child(node_pcdata).set_value(item.c_str());
				}
			}
		}
	} else {
		std::cout << "actions not found" << std::endl;
	}

	field = doc.select_node("//td[@id='expected']");
	if (field) {
	    field.node().append_child(node_pcdata).set_value(database.getExpectedResultsText(id).c_str());
	} else {
		std::cout << "expected not found" << std::endl;
	}

	field = doc.select_node("//td[@id='results']");
	if (field) {
		status = database.getResultsText(id);
	    field.node().append_child(node_pcdata).set_value(database.getResultsText(id).c_str());
	} else {
		std::cout << "results not found" << std::endl;
	}

	field = doc.select_node("//ul[@id='cleanup']");
	if (field) {
		for (std::string cleanupItem : Utilities::split(database.getCleanupText(id),"/hr")) {
			xml_node listItem = field.node().append_child("li");
			listItem.append_child(node_pcdata).set_value(cleanupItem.c_str());
		}
	} else {
		std::cout << "cleanup not found" << std::endl;
	}

	field = doc.select_node("//td[@id='requirements']");
	if (field) {
	    field.node().append_child(node_pcdata).set_value(database.getRequirementsInThisTest(id).c_str());
	} else {
		std::cout << "requirements not found" << std::endl;
	}

	DOMizedTemplate->writeDocument();
	std::cout << "test case constructor completed" << std::endl;

	delete DOMizedTemplate;
}

TestCase::~TestCase() {
	std::cout << "test case destructor starting" << std::endl;
	if (DOMizedTemplate) delete DOMizedTemplate;
}

const std::string& TestCase::getStatus(void) const{
	return status;
}

File: tools_java/tools/src/tools/TestCase.java
package tools;
import org.w3c.dom.Document;
import org.w3c.dom.Element;

/**
 * 
 * @author broihier
 *<pre>
 *Class for creating TestCase objects
 *</pre>
 */
public class TestCase {
	private Document doc = null;
	private String status = "UNTESTED";
	private DOMizeTemplate DOMizedTemplate;
	/** 
	 * Constructor for creating Test Case objects
	 * 
	 * <pre>
	 * create a DOM using the TEST_CASE_TEMPLATE.html file;
	 * insert the test case ID into the DOM;
	 * insert the title into the DOM;
	 * foreach requirement associated with this test case {
	 *     put the expanded requirement into the objective requirements list of the DOM;
	 * }
	 * put the objective information in the DOM;
	 * put the setup information into the DOM;
	 * put the procedure steps (including bullets) into the DOM;
	 * put the expected results into the DOM;
	 * put the results into the DOM;
	 * set the test case status to the value of the results;
	 * put the cleanup steps into the DOM;
	 * put the requirement tags into the DOM;
	 * write the expanded DOM to disk;
	 * </pre>
	 * 
	 * @param id test case ID
	 * @param category test category (chapter/section)
	 * @param database test database for the test category
	 * 
	 */
	public TestCase (String id, String category, TestDatabase database) {
		this.DOMizedTemplate = new DOMizeTemplate("../tools_configuration/TEST_CASE_TEMPLATE.html","./"+category+"/"+id+".html");
		this.doc = DOMizedTemplate.getDoc();
		// insert identification into test case DOM
		Element element = Utilities.findElementById(doc,"testCaseId");
		element.appendChild(doc.createTextNode(id));
		
		element = Utilities.findElementById(doc,"title");
		element.appendChild(doc.createTextNode(database.getIdentificationText(id)));
		
		// insert objective into the test case DOM
		element = Utilities.findElementById(doc,"requirementsList");
		for (String requirement : database.getRequirementsInThisTest(id).replaceAll(" ", "").split(",")){
			Element listItemElement = doc.createElement("li");
			listItemElement.appendChild(doc.createTextNode(database.getRequirementsText(requirement)));
			element.appendChild(listItemElement);
		}
		Element paragraphElement = doc.createElement("p");
		paragraphElement.appendChild(doc.createTextNode("This test case is intended to verify the following implied or specified requirements:"));
		element = Utilities.findElementById(doc,"objective");
		element.insertBefore(paragraphElement, element.getFirstChild());
		paragraphElement = doc.createElement("p");
		paragraphElement.appendChild(doc.createTextNode(database.getObjectiveText(id)));
		element.appendChild(paragraphElement);
		
		// insert setup into the test case DOM
		element = Utilities.findElementById(doc,"setup");
		for (String setup : database.getSetupText(id).split("/hr")) {
			Element listItem = doc.createElement("li");
			listItem.appendChild(doc.createTextNode(setup));
			element.appendChild(listItem);
		}
		
		// insert procedure steps into the test case DOM
		element = Utilities.findElementById(doc,"actions");
		for (String step : database.getProcedureText(id).split("/step")){
			Element listItem = doc.createElement("li");
			Element unOrderedList = doc.createElement("ul");
			if (step.equals("")) continue;
			boolean procedureStep = true;
			for (String bulletText : step.split("/bullet")) {
				if (procedureStep) {
					// this is the procedure step - add it in
					listItem.appendChild(doc.createTextNode(bulletText.replace("\\step", "")));
					procedureStep = false;
				} else {
					Element bulletListItem = doc.createElement("li");
					bulletListItem.appendChild(doc.createTextNode(bulletText.replace("\\bullet","").replace("\\step", "")));
					unOrderedList.appendChild(bulletListItem);
				}
			}
			if (unOrderedList.hasChildNodes()) {
				unOrderedList.setAttribute("type", "square");
				listItem.appendChild(unOrderedList);
			}
			element.appendChild(listItem);
		}
		
		// insert expected results into the test case DOM
		element = Utilities.findElementById(doc,"expected");
		element.appendChild(doc.createTextNode(database.getExpectedResultsText(id)));
		
		// insert results into the test case DOM
		element = Utilities.findElementById(doc,"results");
		this.status = database.getResultsText(id);
		element.appendChild(doc.createTextNode(this.status));
		
		// insert cleanup into the test case DOM
		element = Utilities.findElementById(doc,"cleanup");
		for (String setup : database.getCleanupText(id).split("/hr")) {
			Element listItem = doc.createElement("li");
			listItem.appendChild(doc.createTextNode(setup));
			element.appendChild(listItem);
		}
		
		// insert reauirement IDs into the test case DOM
		element = Utilities.findElementById(doc,"requirements");
		element.appendChild(doc.createTextNode(database.getRequirementsInThisTest(id)));

		// write updated DOM
		DOMizedTemplate.writeDocument();
	}
	/**
	 * Method to return the status of a test case
	 * @return status - getter
	 */
	public String getStatus() {
		return status;
	}

}
